<a href="https://colab.research.google.com/github/isnasabrinaiftinan/praktikum-ipsd/blob/main/Pengolahan_Praproses_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IPSD — Minggu 2: Pengolahan/Praproses Data

**Mata Kuliah:** Infrastruktur dan Platform untuk Sains Data  
**Sub-CLO:** Sub-CLO-09-2-1  
**AI Level:** L2  

**AI Level 2 — AI-Assisted Idea Generation and Structuring.** AI boleh digunakan untuk brainstorming, decomposition, alternatif struktur/pipeline, dan pertanyaan klarifikasi. Jawaban akhir, kode akhir, keputusan, dan interpretasi harus dibuat mahasiswa. Simpan prompt log.

## Capaian Praktikum
- Mengidentifikasi masalah kualitas data.
- Melakukan penanganan missing value, duplicate, tipe data, encoding, scaling, dan outlier sederhana.
- Menyusun preprocessing pipeline yang dapat dijelaskan.

**Dataset:** Breast Cancer Wisconsin dari scikit-learn, dimodifikasi dengan missing value dan duplikasi sintetis.

## Alur Praktikum
- 1. Load dataset built-in scikit-learn.
- 2. Sisipkan missing value dan duplikasi untuk latihan.
- 3. Audit kualitas data dengan `info()`, `isna()`, `duplicated()`, dan statistik deskriptif.
- 4. Lakukan imputasi, hapus duplikasi, dan scaling pada fitur numerik.

In [2]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

data = load_breast_cancer(as_frame=True)
df = data.frame.copy()
# Sisipkan masalah kualitas data untuk latihan
rng = np.random.default_rng(42)
idx = rng.choice(df.index, size=15, replace=False)
df.loc[idx, 'mean radius'] = np.nan
df = pd.concat([df, df.iloc[:3]], ignore_index=True)
print(df.shape)
print('Missing:', int(df.isna().sum().sum()))
print('Duplicate:', int(df.duplicated().sum()))

# Contoh cleaning dasar
df = df.drop_duplicates().reset_index(drop=True)
imputer = SimpleImputer(strategy='median')
df[['mean radius']] = imputer.fit_transform(df[['mean radius']])
print('Sesudah cleaning:', df.shape, int(df.isna().sum().sum()))

(572, 31)
Missing: 15
Duplicate: 3
Sesudah cleaning: (569, 31) 0


## Latihan Terstruktur
- Bandingkan minimal dua strategi imputasi pada satu fitur.
- Tentukan kapan scaling diperlukan dan kapan tidak.
- Buat tabel before/after berisi jumlah missing, duplicate, dan rentang nilai fitur.

In [4]:
# TODO: Kerjakan latihan pada sel ini dan tambahkan sel baru bila diperlukan.

nilai_terbesar = df['mean radius'].max()
nilai_terkecil = df['mean radius'].min()
rata_rata_nilai = df['mean radius'].mean()
nilai_tengah = df['mean radius'].median()

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

hasil = df.copy()
hasil['mean radius'] = scaler.fit_transform(df[['mean radius']])
print(hasil[['mean radius']].head())
print(hasil['mean radius'].min())
print(hasil['mean radius'].max())



   mean radius
0     1.115528
1     1.856558
2     1.603803
3    -0.771514
4     1.776136
-2.0464873171537477
4.022203201208941


In [6]:
print("Min Awal:", df['mean radius'].min())
print("Max Awal:", df['mean radius'].max())

Min Awal: 6.981
Max Awal: 28.11


## Luaran
Notebook + tabel before/after + prompt log L2.

**Bukti AI/validasi:** Prompt log; alasan memilih preprocessing.

##tabel before/after
| No | Kondisi | Jumlah Missing | Jumlah duplikat| Min | Max |
|---|---|---|---|---|---|
|1|awal / before|15|3|6.981|28.11|
|2|sesudah / after|0|0|-2.046|4.022



## Prompt Log — wajib
Isi minimal 2 interaksi yang benar-benar Anda gunakan.

| No | Tujuan Prompt | Prompt | Ringkasan Respons AI | Keputusan Saya / Apa yang Dipakai | Verifikasi |
|---|---|---|---|---|---|
|1|mendapat penjelasan terkait tugas|"hai aku ada tugas ipsd week 2, tolong kamu jelaskan."|Berikut adalah penjelasan konsep dan panduan terstruktur untuk menyelesaikan latihanmu:1. Memahami Alur Latihan TerstrukturBerdasarkan instruksi di notebook, ada 3 tugas utama yang harus kamu selesaikan di bagian latihan:A. Membandingkan Minimal Dua Strategi Imputasi pada Satu Fitur [1]|saya membaca dan memahami|memahami|
|2|menanyakan cara membuat min max|"seingatku kalau cari max for i !=0 and i > n {return max}"|Logika berpikir kamu sudah punya konsep yang bagus tentang membandingkan angka satu per satu! Kode yang kamu tulis itu adalah bentuk algoritma dasar menggunakan perulangan (looping).Dalam algoritma dasar, cara mencari nilai maksimal biasanya kita membuat satu variabel "penampung" nilai tertinggi sementara, lalu kita cek semua angka di dalam data kita. |saya memperbaiki kode yang saya buat di awal menggunakan bantuan ai dengan bahasa python|menulis kembali kode dan coba run|
|3|masih dengan butuh penjelasan|"sebentar, aku masih bingung harus buat apalagi selanjutnya. aku paham kok penjelasan kamu, tapi aku bingung harus buat nya bagaimana."|ai menjelaskan ulang dan memberitahukan apa yang harus dikerjakan|memahami|memahami|
|4|mengecek apakah logika sudah benar atau tidak|"gini ga buat awalnya (nilai_terbesar = df['mean radius'].max() nilai_terkecil = df['mean radius'].min() rata_rata_nilai = df['mean radius'].mean() def StandardScaler(df)"|ai menjelaskan bahwa kode itu sudah benar tapi bagian def masih salah|memperbaiki kode|memperbaiki kode|
|5|cek kembali karena masih salah|"gini ga?? nilai_terbesar = df['mean radius'].max()  nilai_terkecil = df['mean radius'].min()  rata_rata_nilai = df['mean radius'].mean()   nilai_tengah = df['mean radius'].median()  from sklearn.preprocessing import StandardScaler scaler = StandardScaler()  hasil = pd.DataFrame"|Iya, keren banget Isna! Logika susunan kodemu sudah makin matang dan mengarah ke jalur yang benar.Baris 1 sampai 4 untuk mengambil nilai statistik awal (max(), min(), mean(), dan median()) sudah 100% tepat. Pendefinisian scaler = StandardScaler() juga sudah benar.Namun, ada satu bagian di baris terakhir yang perlu sedikit kita sempurnakan, yaitu pada bagian hasil = pd.DataFrame.|memperbaiki bagian pd.DataFrame|memperbaiki kode|
|6|ternyata eror, bantu perbaiki eror|[memberikan kode eror nya]|Error NameError: name 'df' is not defined ini artinya Jupyter Notebook kamu belum mengenal atau lupa apa itu df.Hal ini biasanya terjadi karena kotak kode contoh yang ada di atasnya (kotak tempat data dimuat dan dirusak) belum kamu jalankan (Run) sejak notebook ini dibuka, atau runtime/kernel notebook kamu baru saja ter-restart.|mencoba run cell pertama dulu, baru kemudian run ulang cell kedua|run code|
|7|validasi output|[mengirim screenshot hasil]|ai menyarankan untuk menambah output max dan min|saya menambahkan output max min|run code kembali|

## Refleksi Singkat
1. Bagian apa yang paling terbantu oleh AI?  

= penjelasan mengenai tugas, membuat syntax, memperbaiki eror, dan jawaban dari pertanyaan yang saya tidak pahami

2. Bagian apa yang tetap harus Anda putuskan sendiri?  

= saya memberikan logika max di awal agar syntax yang digunakan tidak susah

3. Apa satu keterbatasan/risiko respons AI yang Anda temukan?

= kadang ai langsung memberikan kode nya, padahal saya masih butuh dijelaskan dulu.